Support Multiple Databases

DanrwAU opened the pull request:
# Description

Complete re-write of database driver to remove native SQLite 3 client and replace with Knex query builder.

Add support for MySQL with possiblity of Amazon RDS and Postgres to be implemented later

Fixes #10 

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

**Need to Enter Tests Completed**

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated changelog.md with changes made

marshyonline commented:
Ohhhhhh me likey!
Ill be happy to move to a nice beefy MySQL server :D
SQL isnt my thing - it breaks my brain too

DanrwAU commented:
Line 126-127 and 130-131 isn't working so far and i haven't found a way around it. 

The where statement inside the join is making it have a hissy fit. 😭

DanrwAU commented:
Well message loading now works and the homepage is accessible! 

I'll work on converting the rest of this as i get time 🍾

davidmckenzie commented:
Wiew 💪 nice one dude :D Will spend some time mentally parsing this soon

davidmckenzie commented on server/public/templates/admin/settings.html:
```diff
@@ -37,25 +37,41 @@ <h5>Global</h5>
                   <option value="sqlite">SQLite</option>
-                  <option value="mongo">MongoDB</option>
-                  <option value="dynamo">DynamoDB</option>
+                  <option value="postgress">Postgress</option>
```
Minor niggle ;) s/Postgress/Postgres/

DanrwAU replied:
Oopsie can't say I've ever used postgres ... Guess my brain autocorrected it lol.

DanrwAU commented:
The homepage loads without errors but still need to confirm the queries are returning the same results as pre conversion.

Looking through the rest of API.js there is still a few statements I'm not sure how to replicate yet so we'll see how it goes :-)

DanrwAU commented:
All the Get's are now done except for message searching - need to figure out how the indexes are going to work with other DB types first i guess. 

Selecting individual capcodes isn't working - not sure what it's problem is. 
Throwing this error in the browser console
```
Error: [$resource:badcfg] http://errors.angularjs.org/1.6.4/$resource/badcfg?p0=get&p1=object&p2=array&p3=GET&p4=%2Fapi%2Fcapcodes%2F916
    at angular.js:38
    at angular-resource.js:758
    at angular.js:16832
    at m.$digest (angular.js:17971)
    at m.$apply (angular.js:18269)
    at l (angular.js:12387)
    at XMLHttpRequest.A.onload (angular.js:12541) "Possibly unhandled rejection: {}"
```
The query Knex is spitting out looks correct so i suspect it's something web related i don't understand :P
```
{ method: 'select',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [ 'new' ],
  __knexQueryUid: '5cae152c-efcc-429b-9dce-f78261e4dbaa',
  sql: 'select * from `capcodes` where `id` = ?' }
```

marshyonline commented:
I'm really liking the way this is taking shape 
I'll pull it down tonight and take a look :)

DanrwAU commented:
The big problem is going to be the REPLACE INTO statements. 
Knex doesn't support these as they are a sqlite thing. Those are the only one's i'm not sure how to handle now 😄 

CapcodeRefresh is definetly broken, not sure what's wrong with it at the moment and individual capcodes won't load onto the aliasdetails screen, not sure there either.

DanrwAU commented:
Couldn't get the capcodeRefresh working, not sure if it's me or Knex at this point.
Knex issue here if anyone wants to keep an eye on it 😉 
https://github.com/tgriesser/knex/issues/2984

Will likely need to be reconsidered anyway as i don't believe REPLACE is valid in MySQL/Postgres anyway?

davidmckenzie commented:
REPLACE is definitely a thing in MySQL and Postgres... I'd have to play with knex but I wonder if it's just shitting the bed because the replace is in the order by.

Something like this might play nicer:

```sql
UPDATE messages 
 SET alias_id = (
   SELECT id FROM (
     SELECT id, REPLACE(address, '_', '%') as addrmatch FROM capcodes 
      WHERE messages.address LIKE address 
      ORDER BY addrmatch DESC LIMIT 1
    )
  );
```

DanrwAU commented:
Interesting, Maybe it's REPLACE INTO i'm thinking of.

Anyway, it seems to me that Knex is escaping the single quotes in the replace which is causing the issue but i am but a simpleton. 

I left my dev laptop at home, i'll have a play with it after work and see if that change helps. 😜

davidmckenzie commented:
May also be worth trying different quoting like:

``.orderByRaw(`REPLACE(address, "_", "%") DESC LIMIT 1`)``

DanrwAU commented:
Yeah i spent an hour messing around with different quoting .. doing it with double quotes or backticks results in an "invalid column name _" error

DanrwAU commented:
Knex team closed my issue as not a bug and provided this as an alternative. 

I'll throw it in tonight and see how it goes 😄 

```
db('messages').update('alias_id', function() {
    this.select('id')
        .from('capcodes')
        .where('address', 'like', knex.ref('messages.address'))
        .orderByRaw("REPLACE(address, '_', '%') DESC LIMIT 1")
  })
```

DanrwAU commented:
Yay! That commit fixed that. 

Knex also seems to have dropped the runtime down to 45 seconds from 67 seconds for me ... don't know how, don't know why 😝

DanrwAU commented:
Almost finished the message post route, Line 581 onwards currently broken.

DanrwAU commented:
Everything in api.js is done except Message searching and capcode adding. '
Knex spits out arrays of objects even for single rows so had to do some hackery around that. 

Let me know if you know a better way than what i've done :-P

DanrwAU commented:
Everything seems to be done except message searching, having a nightmare trying to get the query structure right. 

Current code results in this statement when searching. 
```
sql:
   'select `messages`.*, `capcodes`.`alias`, `capcodes`.`agency`, `capcodes`.`icon`, `capcodes`.`color`, `capcodes`.`ignore`, `capcodes`.`id` from `messages_search_index` left join `messages` on `messages`.`id` = `messages_search_index`.`rowid` left join `capcodes` on `capcodes`.`id` = `messages`.`alias_id` where `messages_search_index` like ? order by `messages`.`timestamp` DESC',
  as: 'aliasMatch' }
```

No results are found on queries that definetly should have a result.. 

@marshyonline @davidmckenzie  Any ideas what's wrong with that query?

davidmckenzie commented:
Yeah that search thing will be a tough one - it's a weird SQLite function that won't translate into cross-platform functionality. May need to make the searching thing raw SQL to match the current query, with a condition to do something similar in other DB engines... will have to look at how FTS is handled in MySQL/Postgres

marshyonline commented:
Yeah SQL isn't for me 😂
Hell of a Datacenter admin but SQL breaks my brain 😂

DanrwAU commented:
> Yeah SQL isn't for me 😂
> Hell of a Datacenter admin but SQL breaks my brain 😂

You and me both, i'm just a lowly windows sysadmin ... coding gives me migraines lol

DanrwAU commented:
> Yeah that search thing will be a tough one - it's a weird SQLite function that won't translate into cross-platform functionality. May need to make the searching thing raw SQL to match the current query, with a condition to do something similar in other DB engines... will have to look at how FTS is handled in MySQL/Postgres

Hmmm ok. 

I might throw it in as a raw just to get us going. 

Out of interest i tried MySQL last night, it doesn't like the way a few queries are structured so we'll need to refactor a some of it anyway for compatibility.

DanrwAU commented:
Converting to Raw results in a different problem....

```
<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0000007C6E55C5C1]
Security context: 0x00d73781e6e1 <JSObject>
    1: /* anonymous */ [000001F8BD023F81] [c:\Development\pagermon\server\routes\api.js:~303] [pc=0000007C6E94F455](this=0x01a54f31ad11 <JSGlobal Object>,rows=0x01f8bd023fb9 <JSArray[101273240]>)
    2: tryCatcher(aka tryCatcher) [00000038CB62D461] [c:\Development\pagermon\server\node_modules\bluebird\js\release\util.js:16] [bytecode=000001257B911EC9 offse...

FATAL ERROR: invalid array length Allocation failed - JavaScript heap out of memory
 1: 00007FF63460121A v8::internal::GCIdleTimeHandler::GCIdleTimeHandler+4810
 2: 00007FF6345DA5B6 node::MakeCallback+4518
 3: 00007FF6345DAFA0 node_module_register+2160
 4: 00007FF63486B3EE v8::internal::FatalProcessOutOfMemory+846
 5: 00007FF63486B31F v8::internal::FatalProcessOutOfMemory+639
 6: 00007FF634DA9304 v8::internal::Heap::MaxHeapGrowingFactor+11476
 7: 00007FF63494A62A v8::internal::Factory::NewFixedArrayWithFiller+234
 8: 00007FF6349B21A8 v8::internal::HashTable<v8::internal::NumberDictionary,v8::internal::NumberDictionaryShape>::EntryToIndex+66792
 9: 00007FF634DCF540 v8::internal::Bitmap::IsClean+30464
10: 0000007C6E55C5C1
```

davidmckenzie commented:
I think that might be due to this bit:

```js
    db.raw(sql, query)
      .then ((rows) => {
       console.log(query)
        if (rows) { 
        rowCount = rows.length
        for (row of rows) {
[..]
          if (pdwMode) {
            if (row.ignore == 0)
              rows.push(row);
```

You're pushing the row back into the array that it already exists in, which then gets processed again as part of the for loop, creating infinite loop. Could either rename one of the `rows` vars, or just reverse the logic so it deletes from `rows` when ignore != 0, etc.

DanrwAU commented:
HAHAHAHAHAHAHAHA  That's great. 

One day ill learn how to code.

DanrwAU commented:
Updated the title post with current status and known issues. 

I need to do a full set of testing comparing with the old version to ensure queries return the same results across versions

DanrwAU commented:
Done some limited testing and identified a few issues this morning. 

Also worked up a set of tests to check for each DB type so i don't lose track fo what i've done. 

Hopefully i'll have some time during the week to knock them out for at least SQLite and restore all original functionality, then start working on Maria/MySQL

DanrwAU commented:
All known issues with SQLite are now fixed. 

Hopefully the testing shows up no more issues and i can move on to MySQL

DanrwAU commented:
SQLite Issue 

Alias Match ID is replacing message ID.  

New Plugin Object
```
[pmon]  2019-01-13 11:01:47  debug : { id: **1458**,
  address: '1234567',
  message: 'Test Message Submission 1',
  source: 'Test',
  timestamp: 1547340105,
  alias_id: 1458,
  alias: 'Test Alias - Site 99',
  agency: 'TEST',
  icon: 'wifi',
  color: 'Green',
  ignore: 0,
  pluginconf:
   { Discord: { enable: false, webhook: '' },
     Pushover:
      { enable: true,
        priority: { value: '1' },
        group: 'urnmxjhdhmaf53keab2pe4x4g9sc3y',
        sound: { value: '' } },
     SMTP: { enable: false, mailto: '' },
     Telegram: { enable: false, chat: '' },
     Twitter: { enable: false, hashtag: '' } },
  pluginData: {} }
[pmon]  2019-01-13 11:01:47  debug : afterMessage done
```
Old Plugin Object
```
[pmon]  2019-01-13 11:01:61  debug : { **id: 230482**,
  address: '1234567',
  message: 'Test Message Submission 1',
  source: 'Test',
  timestamp: 1547340105,
  alias_id: 1458,
  alias: 'Test Alias - Site 99',
  agency: 'TEST',
  icon: 'wifi',
  color: 'Green',
  ignore: 0,
  aliasMatch: 1458,
  pluginconf:
   { Discord: { enable: false, webhook: '' },
     Pushover:
      { enable: true,
        priority: { value: '1' },
        group: 'urnmxjhdhmaf53keab2pe4x4g9sc3y',
        sound: { value: '' } },
     SMTP: { enable: false, mailto: '' },
     Telegram: { enable: false, chat: '' },
     Twitter: { enable: false, hashtag: '' } },
  pluginData: {} }
[pmon]  2019-01-13 11:01:61  debug : afterMessage done
```

DanrwAU commented:
SQLite Issue 

PDW Mode - Source Filtering returns less results.

DanrwAU commented:
Message posting looks to be fixed by https://github.com/pagermon/pagermon/pull/174/commits/553cecb25d102693e0ba566b4b3bd484502d8771
all posting tests completed successfully 😁

DanrwAU commented:
On further investigation - all filters returning MORE results are a result of the alias ignore flag being ironically ignored.

davidmckenzie commented:
Might be type error on the ignore flag?

`.where('capcodes.ignore', '=', '0')`

The ignore column is an int, not sure if that makes a difference with knex

davidmckenzie commented:
Think it's an ordering thing... current code:
```	
UPDATE messages SET alias_id = (SELECT id FROM capcodes WHERE messages.address LIKE address ORDER BY REPLACE(address, '_', '%') DESC LIMIT 1);
```
vs generated:
```
update `messages` set `alias_id` = (select `id` from `capcodes` where `address` like `messages`.`address` order by REPLACE(address, '_', '%') DESC LIMIT 1);
```

And just as I was about to hit comment I see you've found this anyway :D

DanrwAU commented:
Stupid databases 😝

DanrwAU commented:
All tests complete except for Plugins. 

I dont have test environments setup for most of them anymore however the plugin data looks fine and pushover is working correctly - I'm confident they won't be affected. 

If anyone has a **_TEST_** environment you don't care about, i'd appreciate any more feedback on running this version with SQLite 🎉

davidmckenzie commented:
Testing on my local box, raw notes:

* Worked fine on first start (after npm i), schema updated fine
* Front page loading fine
* Searching fails - the messageSearch API endpoint timing out, no errors shown in console:

```
::1 - - [13/Jan/2019:04:48:07 +0000] "GET /api/messageSearch?limit=&page=1&q=rfs HTTP/1.1" - - "http://localhost:3000/?q=rfs" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
::1 - - [13/Jan/2019:04:48:47 +0000] "GET /api/messageSearch?agency=TEST&limit=&page=1 HTTP/1.1" - - "http://localhost:3000/?agency=TEST" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
```

* Adding/updating/refreshing alias works
* Database type field in settings page was empty - current default is `sqlite` but setting in settings page changed to `sqlite3` - may be worth changing sqlite3 to just sqlite for backwards compatibility with the current defaults.
* Smashing it with some test messages all worked good, pushed to browser fine
* Tested filters plugin extremely briefly, worked all good

Tests were with PDW mode and secure mode disabled.

DanrwAU commented:
> Testing on my local box, raw notes:
> * Searching fails - the messageSearch API endpoint timing out, no errors shown in console:

Interesting....  i'll take a peek on a blank DB and see if i can break it on my end. 

> * Database type field in settings page was empty - current default is `sqlite` but setting in settings page changed to `sqlite3` - may be worth changing sqlite3 to just sqlite for backwards compatibility with the current defaults.

Knex requires it to be sqlite3 and i couldn't be bothered putting a reference variable in somewhere lol.
I've added it to the known issues list.

I'm testing with one of my prod instances unfortunately still seems to be some issues with wildcard aliases. 

Aliases are applying to messages after a alias refresh but on posting are returning no results. 
I think the issue is here @ line 584 but can't figure out what it is. 

```
                db.from('capcodes')
                  .select('id', 'ignore')
                  .where('capcodes.address', 'like', address)
                  .orderByRaw("REPLACE(address, '_', '%') DESC LIMIT 1")
                  .then((row) => {
```
For reference this is a full wildcard alias i'm doing all the filtering for this one in reader.js
![image](https://user-images.githubusercontent.com/26653344/51081831-97e81200-174d-11e9-820e-3bb5c8c2d054.png)

davidmckenzie commented:
I think that's the ordering thing again, translates to:

```
select `id`, `ignore` from `capcodes` where `address` like ? order by REPLACE(address, '_', '%') DESC LIMIT 1;
```

Reversing order works, e.g.:

```
select `id`, `ignore` from `capcodes` where '1000014' like `address` order by REPLACE(address, '_', '%') DESC LIMIT 1;
```

The wildcards need to be on the right side of the `LIKE` for sqlite to actually parse them as wildcards, otherwise it would treat them as text

davidmckenzie commented:
Knex doesn't like that though :(

```
[pmon]  2019-01-13 16:01:85  error : select `id`, `ignore` from `capcodes` where `1000014` like 'address' order by REPLACE(address, '_', '%') DESC LIMIT 1 - SQLITE_ERROR: no such column: 1000014
```

DanrwAU commented:
Ok that commit broke it more. Now it matches everything to Alias ID 1 lol.

DanrwAU commented:
When in doubt revert to raw commands lol... That seems to have fixed it

DanrwAU commented:
> * Searching fails - the messageSearch API endpoint timing out, no errors shown in console:

This looks like Knex not serializing correctly. 
It looks like it's creating the tables after running the commands to create the search indexes. 
May need to clean up db.js to execute things in order rather than all over the place 🤷‍♂️ 
Restarting the app fixes it after first run.

DanrwAU commented:
MySQL is still broken. 

Issues so far. 
- Foreign Keys cannot link different table types 
    - Modified messages.alias_id to be an integer 
    - Modified capcodes.id to be an integer 
        - This has the flow on of breaking Auto_Increment (only for MySQL, SQLite seems to handle OK)
        - Created Trigger for MySQL to increment the id in the capcodes table on insert 
            - This is not returning the value set to Knex - resulting in the add page redireccting back to the new capcode screen, also not updating the capcode for messages already containing the alias. 

- Capcode Screen is not ordering correctly, suspect MySQL doesn't like this bit 
`order by REPLACE(address, \'_\', \'%\')`

DanrwAU commented:
> > Testing on my local box, raw notes:
> > 
> > * Searching fails - the messageSearch API endpoint timing out, no errors shown in console:
> 
> Interesting.... i'll take a peek on a blank DB and see if i can break it on my end.

@davidmckenzie Found another cause for this ... and fixed it. Hopefully this shouldn't happen anymore.

DanrwAU commented:
Capcode screen is ordering correctly, by address, my mistake.

Interim solution for Capcode saving for MySQL/MariaDB has been added this really really really really needs to be fixed at some stage 😉

DanrwAU commented:
Kinda got MySQL Searching working.. 
Now it's spitting out this wierdness *sigh* 

```
[ RowDataPacket {
    id: 1,
    address: '1234567',
    message: 'Test Message Woooo Hi',
    source: 'Test',
    timestamp: 1547469329,
    alias_id: null,
    alias: null,
    agency: null,
    icon: null,
    color: null,
    ignore: null,
    aliasMatch: null } ]
[ FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'messages',
    orgTable: 'messages',
    name: 'id',
    orgName: 'id',
    charsetNr: 63,
    length: 10,
    type: 3,
    flags: 16939,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'messages',
    orgTable: 'messages',
    name: 'address',
    orgName: 'address',
    charsetNr: 33,
    length: 765,
    type: 253,
    flags: 20489,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'messages',
    orgTable: 'messages',
    name: 'message',
    orgName: 'message',
    charsetNr: 33,
    length: 196605,
    type: 252,
    flags: 20505,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'messages',
    orgTable: 'messages',
    name: 'source',
    orgName: 'source',
    charsetNr: 33,
    length: 196605,
    type: 252,
    flags: 20497,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'messages',
    orgTable: 'messages',
    name: 'timestamp',
    orgName: 'timestamp',
    charsetNr: 63,
    length: 11,
    type: 3,
    flags: 16392,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'messages',
    orgTable: 'messages',
    name: 'alias_id',
    orgName: 'alias_id',
    charsetNr: 63,
    length: 11,
    type: 3,
    flags: 16392,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'capcodes',
    orgTable: 'capcodes',
    name: 'alias',
    orgName: 'alias',
    charsetNr: 33,
    length: 196605,
    type: 252,
    flags: 20504,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'capcodes',
    orgTable: 'capcodes',
    name: 'agency',
    orgName: 'agency',
    charsetNr: 33,
    length: 196605,
    type: 252,
    flags: 16400,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'capcodes',
    orgTable: 'capcodes',
    name: 'icon',
    orgName: 'icon',
    charsetNr: 33,
    length: 196605,
    type: 252,
    flags: 16,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'capcodes',
    orgTable: 'capcodes',
    name: 'color',
    orgName: 'color',
    charsetNr: 33,
    length: 196605,
    type: 252,
    flags: 16,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'capcodes',
    orgTable: 'capcodes',
    name: 'ignore',
    orgName: 'ignore',
    charsetNr: 63,
    length: 11,
    type: 3,
    flags: 0,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true },
  FieldPacket {
    catalog: 'def',
    db: 'pagermon',
    table: 'capcodes',
    orgTable: 'capcodes',
    name: 'aliasMatch',
    orgName: 'id',
    charsetNr: 63,
    length: 11,
    type: 3,
    flags: 20490,
    decimals: 0,
    default: undefined,
    zeroFill: false,
    protocol41: true } ]
```

DanrwAU commented:
Fixed that - Text Searching now works - Address and Agency based searching still broken 😭

DanrwAU commented:
Pushed some additional bugfixes tonight and it's looking good. 
Planning to do some full testing on the SQLite and MySQL Versions tomorrow. 
I'm considering that being enough for this version (hence removing the tests from the PR description), additional db versions can then be added in later.

DanrwAU commented:
@marshyonline @davidmckenzie I've tested this and haven't run into any problems with MySQL and SQLite. 

Would you mind updating some dev environments and see how you go? 

The only things left are the DB Migration files to take over the current user_version stuff in db.js. It can probably stay as is for initial release, just won't be able to make schema changes until sorted. 

I plan on setting up another Pseudo production environment tonight to do some performance testing, but side by side on my desktop MySQL is a little slower

marshyonline commented:
Ohh nice!
Ill try find some time this week to throw some tests together

DanrwAU commented:
Still need to figure out DB Upgrades for MySQL - preferably removing most of db.js and using knex migration files instead.

DanrwAU commented:
Converted to Knex migration files for DB Management. 
This hasn't been tested at all apart from an initial db seed. Strongly strongly advise never to run the current build unless you know what your doing ;-)

DanrwAU commented:
@pagermon/collaborators mind casting an eye over this one, i think its done 🥇

davidmckenzie commented:
Ooh nice, will have to find some time to test!

davidmckenzie commented on CHANGELOG.md:
```diff
@@ -1,5 +1,12 @@
 
+* Compeletly rebuilt database driver to use the [KnexJS](https://knexjs.org/) query builder. #174
+    * **Upgrades of existing SQLite databases must be performed from v0.2.3**
+    * Add's support for **MySQL**.
```
```suggestion
    * Adds support for **MySQL**.
```

DanrwAU replied:
I need to stop doing this stuff in the middle of the night lol

marshyonline commented:
I shall try and find some time this week to test :)

DanrwAU commented:
Thanks Guys, I'm still finding little quirks here and there but on the whole seems to be working fairly well. 

If you use MySQL - let me know if your getting any Connection Reset errors in your logs, i suspect it's an azure thing for me.

DanrwAU commented:
@davidmckenzie might need your help with this one. 

Testing the capcodeRefresh route has unveiled that no matter how beefy the MySQL Server it is very slow on a large database. 

Currently seeing results like this updateMap: 368070.940ms with 350k messages and 2k aliases. 

Any thoughts on how we can optimize this process?

davidmckenzie commented:
Might need to do some digging on that one... do you get the same results when you run the equivalent SQL directly in the DB? Knex has a thing that prints out the actual SQL it's executing right? Does it differ from the old query at all? And if so, does the old query work magically on MySQL or same problem?

Think this thing is always gonna be a problem with scaling on any RDBMS, some workarounds might be to restrict to only the last X messages, run it as a scheduled job at regular intervals (overnight?), or some combination of both

DanrwAU commented:
Yeah same issue running from the SQL Console directly, the knex string looks perfect compared to the old one. 

I was thinking of having it trigger a job in the background when a change is made for MySQL etc as the database is still useable while it's running as opposed to SQLite where it locks up.

These are all wild pipe dreams i have no idea how to implement tho 🤔

davidmckenzie commented:
Only problem with triggers is they can build up... have seen a few customers do something similar and all is perfect until someone comes along and updates 100 things in a row, and now you got 100 big-ass queries locking up tables for a while 😀 

Could use the refresh-required flag - have a scheduled job that checks for that flag every 10-15 minutes or something, if present it runs the refresh. 

Have used this before in other projects: https://www.npmjs.com/package/cron

Other thing we should probably do is make the refresh route itself asynchronous, so if you trigger a refresh manually, it's not gonna wait until it's done before it comes back to the user. Just return ok and have some kind of feedback when it's actually done or something.

Will have a think, see if I can come up with any wacky ideas

DanrwAU commented:
@davidmckenzie i've started trying to add a CronJob in the latest commit, however i'm dumb and have no idea how to call the AliasRefresh route from there .. Any ideas to nudge me in the right direction?

davidmckenzie commented:
Don't need to call the route or anything, the code's simple enough you can just duplicate it, should be able to just copy pasta the route and strip out the routey bits yeah?

DanrwAU commented:
@pagermon/collaborators What's your thoughts on the CronJob, Run for all DB Types or restrict it to MySQL/MariaDB?

My thinking if we leave it on for SQLite is it has the potential to cause a unexplained (to the admin at least) lockup while it's doing the refresh. 

Thoughts?

davidmckenzie commented:
Yep probably a good call, just a workaround until we can figure out a smarter way to do it anyway :D

davidmckenzie commented:
Still haven't had a chance to set up a test env for this :( Possibly this weekend, if not I got a week off for school hols coming up where I should be able to get some stuff done

DanrwAU commented:
That'd be good, I've been testing myself but fresh set of eyes always finds something obvious 🤣

davidmckenzie commented:
Just pulled down latest on my prod box:

```
4|pagermon | 2019-04-25 12:10 +00:00: [pmon]  2019-04-25 12:10:00  info : Listening on port 3001
4|pagermon | 2019-04-25 12:10 +00:00: [pmon]  2019-04-25 12:10:00  info : Current DB version: 20181118
4|pagermon | 2019-04-25 12:10 +00:00: [pmon]  2019-04-25 12:10:00  info : Checking for database upgrades
4|pagermon | 2019-04-25 12:10 +00:00: Use async .hasTable to check if table exists and then use plain .createTable. Since .createTableIfNotExists actually just generates plain "CREATE TABLE IF NOT EXIST..." query it will not work correctly if there are any alter table queries generated for columns afterwards. To not break old migrations this function is left untouched for now, but it should not be used when writing new code and it is removed from documentation.
4|pagermon | 2019-04-25 12:10 +00:00: migration file "20190322204646_create_capcodes_table.js" failed
4|pagermon | 2019-04-25 12:10 +00:00: migration failed with error: create unique index `cc_pk_idx` on `capcodes` (`id`, `address`) - SQLITE_ERROR: index cc_pk_idx already exists
4|pagermon | 2019-04-25 12:10 +00:00: [pmon]  2019-04-25 12:10:01  error : create unique index `cc_pk_idx` on `capcodes` (`id`, `address`) - SQLITE_ERROR: index cc_pk_idx already exists
```

Can that fail softly? Might need to drop and recreate index on startup

davidmckenzie commented:
```
[STREAMING] Now streaming realtime logs for [all] processes
4|pagermon | 2019-04-25 22:59 +00:00: [pmon]  2019-04-25 22:59:45  info : Listening on port 3001
4|pagermon | 2019-04-25 22:59 +00:00: [pmon]  2019-04-25 22:59:45  info : Current DB version: 20181118
4|pagermon | 2019-04-25 22:59 +00:00: [pmon]  2019-04-25 22:59:45  info : Checking for database upgrades
4|pagermon | 2019-04-25 22:59 +00:00: [pmon]  2019-04-25 22:59:45  info : Setting DB to version: 20190106
```

Looking good 👍

DanrwAU commented:
Only took 62 commits ..... :(

davidmckenzie commented:
Think we missed a console.time somewhere:

```
4|pagermon | 2019-04-26 00:10 +00:00: init: 0.630ms
4|pagermon | 2019-04-26 00:10 +00:00: (node:10703) Warning: No such label 'init' for console.timeEnd()
4|pagermon | 2019-04-26 00:10 +00:00: sql: 3.452ms
4|pagermon | 2019-04-26 00:10 +00:00: send: 2.204ms
4|pagermon | 2019-04-26 00:10 +00:00: send: 7.059ms
4|pagermon | 2019-04-26 00:10 +00:00: (node:10703) Warning: No such label 'sql' for console.timeEnd()
```

This was when navigating between pages with the `GET /api/messages` call. For some reason it's only happening sometimes.

davidmckenzie commented:
Also, tested #260 on the knex branch, so importing and alias refreshing after big updates working good 💯

DanrwAU commented:
> Think we missed a console.time somewhere:
> 
> ```
> 4|pagermon | 2019-04-26 00:10 +00:00: init: 0.630ms
> 4|pagermon | 2019-04-26 00:10 +00:00: (node:10703) Warning: No such label 'init' for console.timeEnd()
> 4|pagermon | 2019-04-26 00:10 +00:00: sql: 3.452ms
> 4|pagermon | 2019-04-26 00:10 +00:00: send: 2.204ms
> 4|pagermon | 2019-04-26 00:10 +00:00: send: 7.059ms
> 4|pagermon | 2019-04-26 00:10 +00:00: (node:10703) Warning: No such label 'sql' for console.timeEnd()
> ```
> This was when navigating between pages with the `GET /api/messages` call. For some reason it's only happening sometimes.

Yup i've seen this too, as far as i can tell all the Console.time's exist not sure what causes it. I've seen it on 0.2.3 as well

DanrwAU commented:
I think this is OK to merge into 0.3.0 now, i'll do finally testing on the full version and bugfix anything else there. 😁

davidmckenzie commented:
Merge and release whenever you're ready ;)
